### DISCUSSION TITLE
Configuring for private Go modules on github.com

### DISCUSSION THREAD

**Discussion by jault3** [renovatebot/renovate#16006] [Request Help] [answered]
Hello 👋 I am using the [Renovate Github App](https://github.com/apps/renovate) and trying to get renovate to work with my Go repos. One repo I am having issues with has a separate private repo from the same organization as a dependency. I cannot figure out how to configure renovate to successfully authenticate and fetch this module when running `go mod tidy`.

Since I am _not_ self hosting, I did not see any way to set the `GOPRIVATE` environment variable. So I think the proper way to do this is with github.com is via `hostRules`. According to [this page](https://docs.renovatebot.com/getting-started/private-packages/#gomod), I believe I need a github.com token in my configuration. But based on [this comment](https://github.com/renovatebot/renovate/issues/3202#issuecomment-480736610) I would expect renovate to reuse the token from installing the app and not require my to put it in plain text or encrypted. Is there a specific format I need to specify this in my config? Here's what I have below:

```
{
  "extends": [
    "config:base"
  ],
  "postUpdateOptions": [
    "gomodTidy",
    "gomodUpdateImportPaths"
  ],
  "hostRules": [
    {
      "matchHost": "https://github.com/MYORG/",
      "hostType": "github",
      "token": "$GITHUB_TOKEN"
    }
  ]
}
```

And just for clarification, `$GITHUB_TOKEN` is not a placeholder for the actual value I put in that file. That is the literal string I entered. The error I receive is:

```
fatal: could not read Password for 'https://**redacted**@github.com': terminal prompts disabled
```

I have tried various combinations of matchHost and token, such as `github.com`, `github.com/MYORG`, `https://github.com`, etc. And for token I have tried an empty string, `x-access-token:$GITHUB_TOKEN`, the word `token`, and a few others.

The [golang language page](https://docs.renovatebot.com/golang/#custom-registry-support-and-authentication) seems to only give an example for how to configure the app when self-hosted. I also tried looking through the code a bit to figure out what it needs, and I did end up [here](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/gomod/artifacts.ts#L59-L64), which brought me [to this file](https://github.com/renovatebot/renovate/blob/cefaa7f2d280d12f5556a491f1d794e80747d800/lib/util/git/auth.ts), but unfortunately wasn't able to figure it out after lots of trial and error.

The most relevant other discussion I found was https://github.com/renovatebot/renovate/discussions/13710, which didn't help me resolve this after reading through it. Thanks in advance!

**Comment by rarkins** [comment, marked as answer]
Are both repos you test with private?

I would have expected this to work out of the box, because we already have some code for this, such as here: https://github.com/renovatebot/renovate/blob/c2fc9312da10e638c54d911bbbc72aedd0fd26a8/lib/modules/manager/gomod/artifacts.ts#L27-L41

Maybe as a first step we should add some debugging messages to confirm when we are setting environment secrets for `go`.

You shouldn't need to - and it's not recommended to - configure your own `hostRules` with `hostType=github` because this has the potential to interfere with Renovate itself when it's doing things like creating PRs.

**Reply by jault3** [reply]
That's correct, both repos are private. I tried removing the hostRules, with this as my resulting renovate.json file
```
{
  "extends": [
    "config:base"
  ],
  "postUpdateOptions": [
    "gomodTidy",
    "gomodUpdateImportPaths"
  ]
}
```
then in the logs it looks like it's trying to lookup the dependency on proxy.golang.org and ultimately ends with this error
```
remote: Repository not found.
```

**Reply by rarkins** [reply]
Does this require something like GOPRIVATE to be set in order for go to work correctly?

**Reply by jault3** [reply]
Yeah that's usually what fixes it. GOPRIVATE tells go tools to skip looking up certain modules in the public proxy. You would set GOPRIVATE to something like `GOPRIVATE=github.com/MYORG` for all repos under that org (most common I think), or `GOPRIVATE=github.com/MYORG/private-repo` for a single repo. Then as long as the local git config has been set up to use the access token for authentication, it should work. 

It looks like you added the git configuration in this PR https://github.com/renovatebot/renovate/pull/3495/files#diff-4fa9c53d0ae4207a652d9aec9bd77effeae549d0fe219cf30d543ba4ae991692R71-R75, and I believe it moved to this file now https://github.com/renovatebot/renovate/blob/64f5bd5f03f977ee860a13cb77d0cb52b73380e6/lib/util/git/auth.ts, so I think that part is ok.

**Reply by rarkins** [reply]
Any possibility that would mess up go for use cases where they have *public* packages/repos which we then mark as private? Or does it just mean go will source them direct instead of via the proxy and that works fine?

**Reply by jault3** [reply]
It should result in go pulling them direct from source control, rather than using the public proxy. It will still compute checksums and verify that checksum on subsequent builds and authenticate the packages. But it may be a little bit slower than just using the public proxy. You can also disable the proxy by setting [`GOPROXY=direct`](https://proxy.golang.org), which I believe would be equivalent to setting GOPRIVATE in a way that covers public modules.

There may be some edge cases when skipping the proxy though. For example, if I publish version 1.0.0 of module A, app B would pull that module in to use it and compute the checksum. If I then overwrite version 1.0.0 of module A (delete the release and republish with updated code), and then app C pulls in that updated module, it will have a different checksum that app B does. app B would have a broken build, but app C would work. The public proxy wouldn't have that issue since it would store the checksums on its own and broken both builds at once. But for repos that are internal to that organization I would think thats less of a concern.

**Reply by jault3** [reply]
Hi @rarkins - Is this something that the github app version would support eventually?

**Reply by rarkins** [reply]
In theory it could be done, but I don't have a plan. Do you think GOPRIVATE is something we could automatically detect and set (e.g. always set it for the current org name github.com/foo) or only safe to let the user specify?

**Reply by jault3** [reply]
I think it would be fine to set for the org automatically. I am also looking at self hosting in the mean time to achieve this functionality.

**Reply by jault3** [reply]
Hi @rarkins just wanted to follow up on this. I was actually able to fix this and get things working! It turns out that the renovate app did not have access to the second private module because we were restricting access to a small set of repos during our POC of renovate. Adding the other private module to the list of repos renovate was allowed to see fixed it right away 😬 Sorry for any confusion! That was definitely an oversight on my part.

But I did go down the route of self hosting and learned quite a bit how renovate works. The `GIT_CONFIG_*` environment variables that are set by default, like you mentioned initially, are actually perfect and authorize the app to pull down any repo in the org (as long as the GH app is configured correctly of course 😄 ). I also noticed the `GOPROXY` variable was set to `https://proxy.golang.org,direct`, which I think handles the case where private modules are not cached in the public proxy and falls back to pulling them straight from source. I can't say for certain without further investigation, but I don't believe GOPRIVATE would even be needed because of that. Thanks for all your help looking into this!
